examples/example-xds/src/main/java/io/grpc/examples/helloworldxds/XdsHelloWorldClient.java

```diff
@@ -0,0 +1,107 @@
+
+import io.grpc.Channel;
+import io.grpc.ChannelCredentials;
+import io.grpc.Grpc;
+import io.grpc.InsecureChannelCredentials;
+import io.grpc.ManagedChannel;
+import io.grpc.StatusRuntimeException;
+import io.grpc.examples.helloworld.GreeterGrpc;
+import io.grpc.examples.helloworld.HelloReply;
+import io.grpc.examples.helloworld.HelloRequest;
+import io.grpc.xds.XdsChannelCredentials;
+import java.util.Arrays;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * A simple xDS client that requests a greeting from {@code HelloWorldServer} or {@link
+ * XdsHelloWorldServer}.
+ */
+public class XdsHelloWorldClient {
+  private static final Logger logger = Logger.getLogger(XdsHelloWorldClient.class.getName());
+
+  private final GreeterGrpc.GreeterBlockingStub blockingStub;
+
+  /** Construct client for accessing HelloWorld server using the existing channel. */
+  public XdsHelloWorldClient(Channel channel) {
+    blockingStub = GreeterGrpc.newBlockingStub(channel);
+  }
+
+  /** Say hello to server. */
+  public void greet(String name) {
+    logger.info("Will try to greet " + name + " ...");
+    HelloRequest request = HelloRequest.newBuilder().setName(name).build();
+    HelloReply response;
+    try {
+      response = blockingStub.sayHello(request);
+    } catch (StatusRuntimeException e) {
+      logger.log(Level.WARNING, "RPC failed: {0}", e.getStatus());
+      return;
+    }
+    logger.info("Greeting: " + response.getMessage());
+  }
+
+  /**
+   * Greet server. If provided, the first element of {@code args} is the name to use in the
+   * greeting. The second argument is the target server. A {@code --secure} flag is also accepted.
+   */
+  public static void main(String[] args) throws Exception {
+    String user = "xds world";
+    // The example defaults to the same behavior as the hello world example. To enable xDS, pass an
+    // "xds:"-prefixed string as the target.
+    String target = "localhost:50051";
+    ChannelCredentials credentials = InsecureChannelCredentials.create();
+    if (args.length > 0) {
+      if ("--help".equals(args[0])) {
+        System.out.println("Usage: [--secure] [NAME [TARGET]]\n");
+        System.err.println("  --secure  Use credentials provided by xDS. Defaults to insecure");
+        System.err.println("  NAME      The name you wish to be greeted by. Defaults to " + user);
+        System.err.println("  TARGET    The server to connect to. Defaults to " + target);
+        System.exit(1);
+      } else if ("--secure".equals(args[0])) {
+        // The xDS credentials use the security configured by the xDS server when available. When
+        // xDS is not used or when xDS does not provide security configuration, the xDS credentials
+        // fall back to other credentials (in this case, InsecureChannelCredentials).
+        credentials = XdsChannelCredentials.create(InsecureChannelCredentials.create());
```

Comment:
nit: this can also be
```
credentials = XdsChannelCredentials.create(credentials);
```
to reuse the previous `InsecureChannelCredentials` instance and may be more expressive?

Reply:
> to reuse the previous InsecureChannelCredentials instance

That is an optimization that reduces readability. If creating insecure credentials was a pain, I'd consider for a different sort of clarity. But it is so trivial to create them, it is much more obvious what the value of `credentials` will be here without studying the data flow.

> may be more expressive?

In what way? It seems to be just another indirection.

Reply:
> > may be more expressive?
> 
> In what way? It seems to be just another indirection.

Basically `XdsChannelCredentials.create()` can accept any `ChannelCredentials` as fallback creds. So imagine multiple ways to enter the block:
```
{
  ...
  credentials = XdsChannelCredentials.create(credentials);
  ...
}
```
and `credentials` could be either an `InsecureChannelCredentials` or a `TlsChannelCredentials` based on how it enters this block. So the above syntax expresses that possibility better than
```
XdsChannelCredentials.create(InsecureChannelCredentials.create())
```
But this is a minor point and I am okay with the current syntax too.

Reply:
Oh, I see. Well, it doesn't _actually_ express that if only insecure is being passed. But I take the point. I think the comment explains this point fairly clearly. And I don't think we're too serious about believing that all that many people would take the effort to set up TLS and then have it be overridden by xDS.

Note, I'm also a bit anxious about modifying credentials in a chaining way because the parsing could _easily_ become a `for` loop, since we have flags now.

Reply:
> ... And I don't think we're too serious about believing that all that many people would take the effort to set up TLS and then have it be overridden by xDS.

Hmmm. So are you saying the most common fallback credentials would be `InsecureChannelCredentials`? I thought there are folks who don't want a plaintext channel at any cost so don't want to use `InsecureChannelCredentials`. Even if there own `TlsChannelCredentials` end up not working they can be assured that the communication will not fall back to plaintext willy nilly (also there is no `FailingCredentials`).

> 
> Note, I'm also a bit anxious about modifying credentials in a chaining way because the parsing could _easily_ become a `for` loop, since we have flags now.

That's a good point. In any case my original suggestion was a minor point.

Reply:
> I thought there are folks who don't want a plaintext channel at any cost so don't want to use InsecureChannelCredentials?

You mention `FailingCredentials` later but still ask this question. `FailingCredentials` is the answer to this question. Yes, they don't exist. It missing doesn't mean users are more likely to use TLS credentials.

Reply:
Although the discussion about `InsecureCreds` `TlsCreds` and `FailingCreds` is important, it is not relevant to this particular comment so I am going to resolve it